topdown/parse_units.go

```diff
@@ -70,43 +69,52 @@ func builtinUnits(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) e
 
 	switch unit {
 	case "m":
-		x.SetFloat64(milli)
+		x.SetFloat64(milliSi)
 	case "":
 		x.SetUint64(none)
 	case "k", "K":
-		x.SetUint64(k)
+		x.SetUint64(kSI)
 	case "ki", "Ki":
 		x.SetUint64(ki)
 	case "M":
-		x.SetUint64(m)
+		x.SetUint64(mSI)
 	case "mi", "Mi":
 		x.SetUint64(mi)
 	case "g", "G":
-		x.SetUint64(g)
+		x.SetUint64(gSI)
 	case "gi", "Gi":
 		x.SetUint64(gi)
 	case "t", "T":
-		x.SetUint64(t)
+		x.SetUint64(tSI)
 	case "ti", "Ti":
 		x.SetUint64(ti)
 	case "p", "P":
-		x.SetUint64(p)
+		x.SetUint64(pSI)
 	case "pi", "Pi":
 		x.SetUint64(pi)
 	case "e", "E":
-		x.SetUint64(e)
+		x.SetUint64(eSI)
 	case "ei", "Ei":
 		x.SetUint64(ei)
 	default:
 		return errUnitNotRecognized(unit)
 	}
 
-	numFloat, ok := new(big.Float).SetString(num)
+	numRat, ok := new(big.Rat).SetString(num)
 	if !ok {
 		return errNumConv
 	}
 
-	numFloat.Mul(numFloat, &x)
+	numRat.Mul(numRat, &x)
+
+	// When using just big.Float, we had floating-point precision
+	// issues because quantities like 0.001 are not exactly representable.
+	// Rationals (such as big.Rat) do not suffer this problem, but are
+	// more expensive to compute with in general.
+	numFloat, ok := new(big.Float).SetString(numRat.FloatString(10))
```

Comment:
Can we shortcut this? `FloatToNumber` is just `f.Text(...)`: https://github.com/open-policy-agent/opa/blob/main/topdown/builtins/builtins.go#L171 -- if `big.Rat` has a similar method, we can forego the big.Float conversion

Reply:
Good suggestion! I rolled that in to the latest version.